### PR TITLE
Fix: updated Y-axes tick logic (fixes #235)

### LINKED ISSUE #235: Y-Axis Ticks Property Issue
**Describe the bug**
We were using Carbon Graph@2.4.0 where the y-axis ticks alignment issue was persisting! We had a workaround which used the rangeRounding and padDomain when set to true for both y1 and y2 axis, the ticks were aligned.
Due to the recent fix in Carbon Graph@2.5.1, it seems to be breaking and wasn't aligning.
Giving the tickCount worked but the ticks property seems to be broken as im unable to get the values when given on the axises.

**Link to CodeSandbox**

[Sandbox](https://codesandbox.io/s/jovial-montalcini-s1ykp?file=/src/data.js ) 

**Expected behavior**
*Ideally the tick values on the y-axes should be displayed as in* (This was in 2.4.0) 
Screenshot 2020-06-12 at 12 30 52 PM

But currently its displaying as 
Screenshot 2020-06-12 at 12 30 35 PM

**Desktop (please complete the following information):**

-   OS: MacOS Mojave
-   Browser : Chrome, Firefox

### PR DESCRIPTION
**Pull request by sdadn**
**What is the purpose of this pull request? (Check any that's applicable)**

-   [ ] Feature
-   [x] Bug fix
-   [ ] Dependency updates
-   [ ] Documentation updates
-   [ ] Build related changes
-   [ ] Changes an existing functionality
-   [ ] Other, please explain:

**Does this introduce a breaking change?**

-   [ ] Yes
-   [x] No

**What changes did you make? (Give an overview)**

This code change fixes the bug described in #235 . The tick generation logic has been updated for improved handling of different scenarios. Rendering of ticks when `allowCalibration` is set to _false_ is fixed and custom tick values for the Y axes can now be used again after the updated in 2.5.1. The new logic now utilizes tick values in the following decreasing priority:

* Custom tick values
* ticksCount property
* default method

**Add any screenshots**

Consumers can now define custom tick values again:
Screen Shot 2020-06-24 at 3 41 12 PM

With `allowCallibration` set to _false_, consumer axis limits are used insted of being rounded up or down:
image

### CODE REVIEW

File: src/main/js/helpers/axis.js [hunk 1]
```diff
@@ -191,37 +191,136 @@ const getAxesScale = (axis, scale, config) => {
         getAxisInfoOrientation(config.axis.x.orientation)
     );
 
-    // If ticksCount is undefined or greater than TICKS_MAXCOUNT
-    // AND if the Y2 is visible, then utilize a default value for
-    // the ticksCount. This is based on the ranges of the Y & Y2 axes.
+    if (hasY2Axis(config.axis)) {
+        if (!utils.isUndefined(config.axis.y.ticks.values)) {
```
Comment (Veera81) [inline src/main/js/helpers/axis.js:195]: You can add a comment here if the y-axis tick values are provided, it will take the priority

File: src/main/js/helpers/axis.js [hunk 2]
```diff
@@ -191,37 +191,136 @@ const getAxesScale = (axis, scale, config) => {
...
+                getAxisTickFormat(config.locale, config.axis.y.ticks.format)
+            );
+            axis.y2 = prepareY2Axis(
+                scale.y2,
+                config.axis.y2.ticks.values,
+                config.height,
+                getAxisTickFormat(config.locale, config.axis.y2.ticks.format)
+            );
+            return axis;
+        }
```
Comment (Veera81) [inline src/main/js/helpers/axis.js:209]: We may update the logic like this

```js
// If y-2 axis is present
if (hasY2Axis(config.axis)) {
     // If tick values are provided, ticks calculated based on provided values
     if (utils.isDefined(config.axis.y.ticks.values) && utils.isDefined(config.axis.y2.ticks.values)) {
     ......................................................
    } 
    // If ticksCount property is provided and it is less than the maximum ticks count
   // tick values calculated based on the ticksCount
   else if (utils.isDefined(config.ticksCount) && config.ticksCount <= constants.TICKSCOUNT_MAXLIMIT) {
     ......................................
   }
      // If ticksCount is undefined or greater than TICKS_MAXCOUNT
      // AND if the Y2 is visible, then utilize a default value for
     // the ticksCount. This is based on the ranges of the Y & Y2 axes.
    else {
      .............................
     }

// If y-2 axis not present and only y-axis present
else {
     // If y axis tick values are provided, ticks calculated based on provided values
     if (utils.isDefined(config.axis.y.ticks.values)) {
     ......................................................
    } 
    // If y-axis ticksCount property is provided and it is less than the maximum ticks count
   // tick values calculated based on the ticksCount
   else if (utils.isDefined(config.ticksCount) && config.ticksCount <= constants.TICKSCOUNT_MAXLIMIT) {
     ......................................
   }
      // If ticksCount is undefined or greater than TICKS_MAXCOUNT
      //  then utilize a default value for
     // the ticksCount. This is based on the ranges of the Y axis.
    else {
      .............................
     }
} 
```

File: src/main/js/helpers/axis.js [hunk 3]
```diff
@@ -329,35 +428,41 @@ const getAverageTicksCount = (rangeY, rangeY2) => {
...
  * @param {number} ticksCount - Number of ticks between the upper and lower limits
+ * @param {boolean} allowCalibration - Whether this property is true or false
  * @returns {(Array)} returns array of values to be used as tick labels
  */
 const generateYAxesTickValues = (
     lowerLimit,
     upperLimit,
-    ticksCount = constants.DEFAULT_TICKSCOUNT
+    ticksCount = constants.DEFAULT_TICKSCOUNT,
+    allowCalibration = true
```
Comment (Veera81) [inline src/main/js/helpers/axis.js:438]: Why we are setting `allowCalibration` value to true by default?

File: src/test/unit/controls/Graph/GraphAxes-spec.js [hunk 4]
```diff
@@ -408,6 +408,232 @@ describe("Graph - Axes", () => {
...
+            );
+            const NumberOfTicks = allYAxisElements[0].childNodes.length;
+            expect(
+                allYAxisElements[0].childNodes[1].querySelector("text")
+                    .textContent
+            ).toBe("2");
+            expect(
+                allYAxisElements[0].childNodes[NumberOfTicks - 1].querySelector(
+                    "text"
+                ).textContent
```
Comment (Veera81) [inline src/test/unit/controls/Graph/GraphAxes-spec.js:440]: We can modify the tests
```js
describe("when Y-2 axis is present", () => {
   describe("cutsom y and y-2 tick values are provided and ticksCount property is not provided", () => {
         it("should calculate y and y-2 axis tick values based on the custom y and y-2 values", () => {
            .........
         }

   describe("cutsom y and y-2 tick values are provided and ticksCount property is provided", () => {
         it("should calculate y and y-2 axis tick values based on the custom y and y-2 values", () => {
            .........
         }

   describe("cutsom y and y-2 tick values are not provided and ticksCount property is provided", () => {
         it("should calculate y and y-2 axis tick values based on the ticksCount property", () => {
            .........
         }

   describe("cutsom y and y-2 tick values are not provided and ticksCount property is not provided", () => {
         it("should calculate y and y-2 axis tick values based on use D3.js to automatically", () => {
            .........
         }
}

describe("when Y-2 axis is not present", () => {
   describe("cutsom y tick values are provided and ticksCount property is not provided", () => {
         it("should calculate y tick values based on the custom y values", () => {
            .........
         }

   describe("cutsom y values are provided and ticksCount property is provided", () => {
         it("should calculate y axis tick values based on the custom y values", () => {
            .........
         }

   describe("cutsom y values are not provided and ticksCount property is provided", () => {
         it("should calculate y axis tick values based on the ticksCount property", () => {
            .........
         }

   describe("cutsom y tick values are not provided and ticksCount property is not provided", () => {
         it("should calculate y axis tick values based on use D3.js to automatically", () => {
            .........
         }
}
```

File: src/main/js/helpers/axis.js [hunk 5]
```diff
@@ -191,37 +191,136 @@ const getAxesScale = (axis, scale, config) => {
...
+                config.ticksCount,
+                config.allowCalibration
+            );
+
+            axis.y = prepareYAxis(
+                scale.y,
+                yTickValues,
+                config.height,
+                getAxisTickFormat(config.locale, config.axis.y.ticks.format)
+            );
```
Comment (AshishMotanamGurunadham) [inline src/main/js/helpers/axis.js:235]: May i know why we are calling same method multiple times?

File: src/main/js/helpers/axis.js (hunk 3 above)
Comment (sdadn) [reply src/main/js/helpers/axis.js:438]: Because allowCalibration is true by default according to the documentation

File: src/main/js/helpers/axis.js [hunk 6]
```diff
@@ -329,35 +428,41 @@ const getAverageTicksCount = (rangeY, rangeY2) => {
...
     const tickValues = [];
 
     // use the d3,js nice function to round off the upper and lower limits
     // to multiples of 2, 5 or 10
-    const [newLowerLimit, newUpperLimit] = d3
-        .scaleLinear()
-        .domain([lowerLimit, upperLimit])
-        .nice()
-        .domain();
+    // let [newLowerLimit, newUpperLimit] = [lowerLimit, upperLimit];
```
Comment (AshishMotanamGurunadham) [inline src/main/js/helpers/axis.js:445]: is this line necessary in here as comment?

File: src/main/js/helpers/axis.js [hunk 7]
```diff
@@ -191,37 +191,136 @@ const getAxesScale = (axis, scale, config) => {
...
+            config.height,
+            getAxisTickFormat(config.locale, config.axis.y.ticks.format)
+        );
+
+        return axis;
+    }
+
+    axis.y = prepareYAxis(
+        scale.y,
+        undefined,
```
[the remark this answers is not in the document]
Comment (sdadn) [reply src/main/js/helpers/axis.js:319]: This is for the case where consumer does not define tick values. Functionally, this is identical to the original logic where `config.axis.y.ticks.values` is undefined.

File: docs/helpers/Ticks.md [hunk 8]
```diff
@@ -107,3 +108,21 @@ var axis = {
...
+
+The tick values are set one of 3 methods, in the following priority:
+
+-   **Consumer provided values**:
+
+    Consumer provided values in `axis.y.ticks.values` and `axis.y2.ticks.values` will take priority if provided.
+
+-   **ticksCount property**
+
+    If no values are provided and the ticksCount property is set, then it will be used to calculate tick values by dividing `upperLimit - lowerLimit` into `ticksCount + 1` equal divisions. If ticksCount is an invalid value, such as greater than `TICKSCOUNT_MAXLIMIT`, then this property will be ignored and the default behavior will be used.
```
[the remark this answers is not in the document]
Comment (sdadn) [reply docs/helpers/Ticks.md:122]: As long as `config.axis.y.ticks.values` is defined, custom ticks will take a priority.

File: src/main/js/helpers/axis.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (sdadn) [reply src/main/js/helpers/axis.js:195]: We can check for y2 ticks, but since we are already checking for the y2 axis, it does not change the behavior very much. Not checking for y2 also allows for the additional situation where a consumer only provides Y ticks and y2 ticks are automatically generated. This way, they can have a starting point for y2 ticks and make any changes necessary to align them with the Y ticks.

File: src/main/js/helpers/axis.js (hunk 2 above)
Comment (sdadn) [reply src/main/js/helpers/axis.js:209]: Is there any advantage to updating the logic this way? The behavior is identical to mine, but as a developer, I would prefer to use only `if` and multiple returns, because I can see that once the condition is true, the function ends, so no need to continue reading if I'm debugging a certain condition.

Comment (Veera81) [reply src/main/js/helpers/axis.js:209]: It is not necessary to follow any particular way. But we should try to write a code which is more readable. And the second thing is if it is known that if only one condition can be true at a time then it is better to use the if-else ladder as it will save time for unnecessary checking of all if the condition

File: src/main/js/helpers/axis.js (hunk 6 above)
Comment (sdadn) [reply src/main/js/helpers/axis.js]: thanks for pointing that out, removed here:
https://github.com/cerner/carbon-graphs/pull/249/commits/3b13c155f716f18e013f983f062069f8baaa8cd2

File: src/main/js/helpers/axis.js [hunk 9]
```diff
@@ -191,37 +191,136 @@ const getAxesScale = (axis, scale, config) => {
...
+            scale.y,
+            config.axis.y.ticks.values,
+            config.height,
+            getAxisTickFormat(config.locale, config.axis.y.ticks.format)
+        );
+        return axis;
+    }
+
+    if (
+        !utils.isUndefined(config.ticksCount) ||
```
[the remark this answers is not in the document]
Comment (sdadn) [reply src/main/js/helpers/axis.js]: Updated here:
https://github.com/cerner/carbon-graphs/pull/249/commits/bfa9764f7c6bb675da030c824e0304bece4c317a

File: src/main/js/helpers/axis.js [hunk 10]
```diff
@@ -191,37 +191,136 @@ const getAxesScale = (axis, scale, config) => {
...
             config.height,
-            getAxisTickFormat(config.locale, config.axis.y2.ticks.format),
-            config.ticksCount
+            getAxisTickFormat(config.locale, config.axis.y2.ticks.format)
         );
+
+        return axis;
     }
+
+    if (!utils.isUndefined(config.axis.y.ticks.values)) {
```
[the remark this answers is not in the document]
Comment (sdadn) [reply src/main/js/helpers/axis.js]: Updated here:
https://github.com/cerner/carbon-graphs/pull/249/commits/bfa9764f7c6bb675da030c824e0304bece4c317a

File: src/main/js/helpers/axis.js [hunk 11]
```diff
@@ -191,37 +191,136 @@ const getAxesScale = (axis, scale, config) => {
...
 
-    if (
-        (utils.isUndefined(config.ticksCount) ||
-            config.ticksCount > constants.TICKSCOUNT_MAXLIMIT) &&
-        hasY2Axis(config.axis)
-    )
-        config.ticksCount = getAverageTicksCount(
+        // If ticksCount is defined
+        if (
+            !utils.isUndefined(config.ticksCount) &&
```
[the remark this answers is not in the document]
Comment (sdadn) [reply src/main/js/helpers/axis.js]: Updated here:
https://github.com/cerner/carbon-graphs/pull/249/commits/bfa9764f7c6bb675da030c824e0304bece4c317a

File: src/main/js/helpers/axis.js (hunk 2 above)
Comment (sdadn) [reply src/main/js/helpers/axis.js]: update here:
https://github.com/cerner/carbon-graphs/pull/249/commits/a805317e11652ca6ea8f9c4dc607f86d9e2caf33

File: src/test/unit/controls/Graph/GraphAxes-spec.js (hunk 4 above)
Comment (sdadn) [reply src/test/unit/controls/Graph/GraphAxes-spec.js]: Updated tests here:
https://github.com/cerner/carbon-graphs/pull/249/commits/f8b9b62e502a9b67a5be44e8677ad6a0c6aae847

File: src/main/js/helpers/axis.js (hunk 1 above)
Comment (sdadn) [reply src/main/js/helpers/axis.js]: updated comments:
https://github.com/cerner/carbon-graphs/pull/249/commits/6533ea2824a5399bd0de7bfab38eb879c14af30f

File: docs/core/Graph.md [hunk 12]
```diff
@@ -48,7 +48,7 @@ Graph is a set of graphs that share a common set of base visualization structure
...
 | showShapes             | boolean  | true                                           | Toggle to show shapes in the line graph                                                                                                                                                 |
 | showVGrid              | boolean  | true                                           | Toggle to show vertical grid                                                                                                                                                            |
-| ticksCount             | number   | 3                                              | Number of ticks to show on the vertical axes, excluding the upper and lower limits, if not provided, the number of ticks will be calculated automatically depending on the Y & Y2 range |
+| ticksCount             | number   | 3-7 depending on the input range               | Number of ticks to show on the vertical axes, excluding the upper and lower limits, if not provided, the number of ticks will be calculated automatically depending on the Y & Y2 range |
```
Comment (Veera81) [inline docs/core/Graph.md:51]: So based on the documentation, the consumer can only provider ticksCount between 3 and 7. Does it include 3 and 7 values as well?

File: docs/helpers/Ticks.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (Veera81) [reply docs/helpers/Ticks.md:122]: Can we add a note about this? If consumer provides both `custom values` and `ticksCount`, custom values takes priority?

So that consumer will know about this behavior

File: src/main/js/helpers/axis.js [hunk 13]
```diff
@@ -191,36 +197,139 @@ const getAxesScale = (axis, scale, config) => {
...
     );
 
-    // If ticksCount is undefined or greater than TICKS_MAXCOUNT
-    // AND if the Y2 is visible, then utilize a default value for
-    // the ticksCount. This is based on the ranges of the Y & Y2 axes.
+    if (hasY2Axis(config.axis)) {
+        // Y and Y2 axes - custom tick values. Takes priority
+        // and ignores ticksCount if it is set. Will not work if only
+        // Y2 ticks are provided.
+        if (utils.isDefined(config.axis.y.ticks.values)) {
```
Comment (Veera81) [inline src/main/js/helpers/axis.js:204]: Probably, Is it safe to verify y2-ticks as well utils.isDefined(config.axis.y2.ticks.values)?

What if the consumer only provided y-ticks values and y-2 ticks values are not defined?

File: docs/core/Graph.md (hunk 12 above)
Comment (sdadn) [reply docs/core/Graph.md:51]: Oh, I mean the default values are from 3-7. Otherwise the acceptable range is form 0-19. I'll probably rephrase it to make it clearer.

File: src/main/js/helpers/axis.js (hunk 13 above)
Comment (sdadn) [reply src/main/js/helpers/axis.js:204]: If only Y ticks are provided for a 2 axis graph, then the Y2 ticks will be automatic, as shown below:

image

This decision was deliberate so that consumers can have a starting point for determine what Y2 values should be used to maintain alignment. There is also the case where the automatic values are aligned if both axes have the same range. In this example, the Y ticks are custom while the Y2 ticks are automatic:
image
